### PR TITLE
Remove inactive entities [15956]

### PR DESCRIPTION
**Pull request by jparisu**
Merge after:
- #175 
- #178
- #172 
- #174

### CODE REVIEW

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

File: test/unittest/Database/DatabaseClearTests.cpp [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+
+        // 2 entities remain: Domain and 1 locator
+        ASSERT_EQ(db.get_entity_ids(EntityKind::DOMAIN, EntityId::all()).size(), 1u);
+        ASSERT_EQ(db.get_entity_ids(EntityKind::LOCATOR, EntityId::all()).size(), 1u);
+    }
+}
+
+TEST(database, clear_inactive_entities_database_complex)
+{
+    // TODO
```
[the remark this answers is not in the document]
Comment (jparisu) [reply test/unittest/Database/DatabaseClearTests.cpp:204]: You really don't like TODOs :stuck_out_tongue_closed_eyes:

File: src/cpp/database/database.cpp [hunk 2]
```diff
@@ -35,10 +35,64 @@ namespace eprosima {
...
+        }
+    }
+}
+
+template<typename E>
+void clear_inactive_entities_from_map_(
+        std::map<EntityId, std::map<EntityId, std::shared_ptr<E>>>& map)
+{
+    // The higher map will not be removed because it holds the domain, so
+    // only internal map should be iterate.
```
Comment (MiguelCompany) [inline src/cpp/database/database.cpp:64]:
```suggestion
    // we should just iterate over the internal map.
```

**Review by MiguelCompany — CHANGES_REQUESTED** [review]
Just a nit

File: src/cpp/database/database.cpp [hunk 3]
```diff
@@ -4827,6 +4827,26 @@ void Database::change_entity_status_of_kind(
...
+
+                // subentities DataReaders
+                for (auto dr_it : participant->data_readers)
+                {
+                    change_entity_status_of_kind(
+                        dr_it.first,
+                        active,
+                        EntityKind::DATAREADER,
+                        participant->domain->id);
+                }
```
Comment (MiguelCompany) [inline src/cpp/database/database.cpp:4849]: I would probably only do this when `active` is `false`.

File: include/fastdds_statistics_backend/StatisticsBackend.hpp [hunk 4]
```diff
@@ -433,6 +433,12 @@ class FASTDDS_STATISTICS_BACKEND_DllAPI StatisticsBackend
     static void load_database(
             const std::string& filename);
 
+    //! Clear all statistics data of all entities.
+    static void clear_statistics_data();
+
+    //! Remove all inactive entities from database.
+    static void clear_inactive_entities();
+
```
Comment (MiguelCompany) [inline include/fastdds_statistics_backend/StatisticsBackend.hpp:441]: These two new API methods need to have tests.

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

File: test/unittest/StatisticsBackend/ClearEntitiesTests.cpp [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+    DatabaseEntityQueue* entity_queue = nullptr;
+    // Data queue, attached to the database
+    DatabaseDataQueue* data_queue = nullptr;
+    // Statistics participant_, that is supposed to receive the callbacks
+    eprosima::fastdds::dds::DomainParticipant statistics_participant;
+    // Listener under tests. Will receive a pointer to statistics_participant
+    StatisticsParticipantListener* participant_listener = nullptr;
+};
+
+// Windows dll do not export ParticipantProxyData class members (private APIs)
```
Comment (MiguelCompany) [inline test/unittest/StatisticsBackend/ClearEntitiesTests.cpp:140]:
```suggestion
// Windows dll does not export ParticipantProxyData / WriterProxyData class members (private APIs)
```

**Review by MiguelCompany — CHANGES_REQUESTED** [review]
Small nit

**Pull request merged**